pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,1077 @@
+        ///
+        /// Withdrawals can NOT be cancelled because they inflict ecomomic damage on the lottery through collator unstaking
+        /// and the user causing this damage must be subjected to economic consequences for doing so
+        ///
+        /// The withdrawal is paid from [`RemainingUnstakingBalance`]
+        /// If this balance is too low to handle the request, another collator is unstaked
+        ///
+        /// # Arguments
+        ///
+        /// * `amount` - the amount of funds to withdraw
+        ///
+        /// # Errors
+        ///
+        /// Returns an error if:
+        /// * `amount` is below the minimum withdraw amount
+        /// * `amount` is larger than the user's total deposit
+        /// * It is too close to the drawing
+        /// * The user has no or not enough active funds
+        /// * There are any arithmetic underflows
+        #[pallet::call_index(1)]
+        #[pallet::weight(<T as Config>::WeightInfo::request_withdraw(Pallet::<T>::total_users(), pallet_parachain_staking::Pallet::<T>::selected_candidates().len() as u32))]
+        pub fn request_withdraw(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
+            let caller = ensure_signed(origin)?;
+
+            ensure!(
+                amount >= Self::min_withdraw(),
+                Error::<T>::WithdrawBelowMinAmount
+            );
+            ensure!(
+                Self::not_in_drawing_freezeout(),
+                Error::<T>::TooCloseToDrawing
+            );
+            let now = <frame_system::Pallet<T>>::block_number();
+            log::debug!("Requesting withdraw of {:?} tokens", amount);
+            // Ensure user has enough funds active and mark them as offboarding (remove from `ActiveFundsPerUser`)
+            ActiveBalancePerUser::<T>::try_mutate_exists(caller.clone(), |maybe_balance| {
+                match maybe_balance {
+                    None => Err(Error::<T>::NoDepositForAccount),
+                    Some(balance) => {
+                        // Withdraw only what's active
+                        ensure!(*balance >= amount, Error::<T>::WithdrawAboveDeposit);
+                        // Mark funds as offboarding
+                        WithdrawalRequestQueue::<T>::mutate(|withdraw_vec| {
+                            withdraw_vec.push(Request {
+                                user: caller.clone(),
+                                block: now,
+                                balance: amount,
+                            })
+                        });
+                        // store reduced balance
+                        *maybe_balance = match balance
+                            .checked_sub(&amount)
+                            .ok_or(Error::<T>::ArithmeticUnderflow)?
+                        {
+                            new_balance if new_balance.is_zero() => {
+                                TotalUsers::<T>::mutate(|users| *users -= 1);
```

Comment:
`-` seems kind of unsafe to me here.

Reply:
Shouldnt be an issue, if this is called, users would be obviously 1 or more

Reply:
TotalUsers is only used as a weight hint anyway, no code logic is impacted by this.

Can't hurt to do it properly tho, changed